src/imputors/knn.jl

```diff
@@ -0,0 +1,86 @@
+"""
+    KNN <: Imputor
+
+Imputation using k-Nearest Neighbor algorithm.
+
+# Keyword Arguments
+* `num_nn::Int`: number of nearest neighbors
+* `dist::MinkowskiMetric`: distance metric suppports by `NearestNeighbors.jl` (Euclidean, Chebyshev, Minkowski and Cityblock)
+* `on_complete::Function`: a function to run when imputation is complete
+
+# Reference
+* Troyanskaya, Olga, et al. "Missing value estimation methods for DNA microarrays." Bioinformatics 17.6 (2001): 520-525.
+"""
+# TODO : Support Categorical Distance (NearestNeighbors.jl support needed)
+struct KNN{M} <: Imputor where M <: NearestNeighbors.MinkowskiMetric
+    num_nn::Int
+    dist::M
+    missing_neighbors_threshold::AbstractFloat
+    context::AbstractContext
+end
+
+function KNN(; num_nn=1, missing_neighbors_threshold=0.5,
+    dist=Euclidean(), context=Context())
+
+    if num_nn < 1
+        throw(ArgumentError("The number of nearset neighbors should be greater than 0"))
+    end
+
+    # to exclude missing value itself
+    KNN(num_nn + 1, dist, missing_neighbors_threshold, context)
+end
+
+function impute!(data::AbstractMatrix{<:Union{T, Missing}}, imp::KNN) where T<:Real
+
+    imp.context() do ctx
+        # transpose to D x N for KDTree
+        transposed = float.(collect(transpose(data)))
```

Comment:
Is the `float` and `collect` really necessary here?

Reply:
If this is really necessary then I recommend doing this in 1 operations:

```
transposed = Matrix{Union{Float64, Missing}}(transpose(data))
```

Reply:
NOTE: If you keep this as just `transpose(data)` then any changes to the `transposed` variable below will propagate to the `data` variable, and avoids some extra allocations.